nvflare/app_common/executors/multi_process_executor.py

```diff
@@ -158,116 +175,103 @@ def _initialize_multi_process(self, fl_ctx: FLContext):
-                handle_conn = self._create_connection(open_port)
+            # send the init data to all the child processes
+            cell.register_request_cb(
+                channel=CellChannel.MULTI_PROCESS_EXECUTOR,
+                topic=CellChannelTopic.EXECUTE_RESULT,
+                cb=self.receive_execute_result,
+            )
+            cell.register_request_cb(
+                channel=CellChannel.MULTI_PROCESS_EXECUTOR,
+                topic=CellChannelTopic.FIRE_EVENT,
+                cb=self._relay_fire_event,
+            )
 
-                self.conn_clients.append(
-                    {CommunicationMetaData.EXE_CONN: exe_conn, CommunicationMetaData.HANDLE_CONN: handle_conn}
+            self.targets = []
+            for i in range(self.num_of_processes):
+                fqcn = FQCN.join([cell.get_fqcn(), str(i)])
+                start = time.time()
+                while not cell.is_cell_reachable(fqcn):
+                    time.sleep(1.0)
+                    if time.time() - start > 60.0:
+                        raise RuntimeError(f"Could not reach the communication cell: {fqcn}")
+                self.targets.append(fqcn)
+            request = new_cell_message(
+                {},
+                {
+                    CommunicationMetaData.FL_CTX: get_serializable_data(fl_ctx),
+                    CommunicationMetaData.COMPONENTS: self.components_conf,
+                    CommunicationMetaData.LOCAL_EXECUTOR: self.executor_id,
+                },
+            )
+            replies = cell.broadcast_request(
+                targets=self.targets,
+                channel=CellChannel.CLIENT_SUB_WORKER_COMMAND,
+                topic=MultiProcessCommandNames.INITIALIZE,
+                request=request,
+            )
+            for name, reply in replies.items():
+                if reply.get_header(MessageHeaderKey.RETURN_CODE) != F3ReturnCode.OK:
+                    self.log_exception(fl_ctx, "error initializing multi_process executor")
+                    raise ValueError(reply.get_header(MessageHeaderKey.ERROR))
+        except BaseException as e:
+            self.log_exception(fl_ctx, f"error initializing multi_process executor: {secure_format_exception(e)}")
+
+    def receive_execute_result(self, request: CellMessage) -> CellMessage:
+        return_data = request.payload
+        with self.engine.new_context() as fl_ctx:
+            fl_ctx.props.update(return_data[CommunicationMetaData.FL_CTX].props)
+            self.execute_result = return_data[CommunicationMetaData.SHAREABLE]
+
+        self.execute_complete = True
+        return F3make_reply(ReturnCode.OK, "", None)
+
+    def _relay_fire_event(self, request: CellMessage) -> CellMessage:
+        data = request.payload
+        with self.engine.new_context() as fl_ctx:
+            event_type = data[CommunicationMetaData.EVENT_TYPE]
+            rank_number = data[CommunicationMetaData.RANK_NUMBER]
+
+            with self.relay_lock:
```

Comment:
Why do we need this lock?

Reply:
This is to protect the event handling sequentially.